tests/fetch/cached_xhr.cpp

```diff
@@ -0,0 +1,56 @@
+#include <string.h>
+#include <stdio.h>
+#include <math.h>
+#include <assert.h>
+#include <emscripten/fetch.h>
+
+int result = 0;
+
+// Fetch file without XHRing.
+void fetchFromIndexedDB()
+{
+  emscripten_fetch_attr_t attr;
+  emscripten_fetch_attr_init(&attr);
+  strcpy(attr.requestMethod, "GET");
+  attr.onsuccess = [](emscripten_fetch_t *fetch) {
+    assert(fetch->numBytes == fetch->totalBytes);
+    assert(fetch->data != 0);
+    printf("Finished downloading %llu bytes\n", fetch->numBytes);
+    emscripten_fetch_close(fetch);
+
+#ifdef REPORT_RESULT
+    result = 1;
+    REPORT_RESULT();
+#endif
+
+  };
+  attr.onprogress = [](emscripten_fetch_t *fetch) {
+    printf("Downloading.. %.2f%% complete.\n", (fetch->dataOffset + fetch->numBytes) * 100.0 / fetch->totalBytes);
+  };
+  attr.attributes = EMSCRIPTEN_FETCH_APPEND | EMSCRIPTEN_FETCH_NO_DOWNLOAD;
+  emscripten_fetch_t *fetch = emscripten_fetch(&attr, "gears.png");
+}
+
+// XHR and store to cache.
+int main()
+{
+  emscripten_fetch_attr_t attr;
+  emscripten_fetch_attr_init(&attr);
+  strcpy(attr.requestMethod, "GET");
+  attr.onsuccess = [](emscripten_fetch_t *fetch) {
+    assert(fetch->numBytes == fetch->totalBytes);
+    assert(fetch->data != 0);
+    printf("Finished downloading %llu bytes\n", fetch->numBytes);
+    emscripten_fetch_close(fetch);
+
+    // Test that the file now exists:
+    fetchFromIndexedDB();
+  };
+  attr.onprogress = [](emscripten_fetch_t *fetch) {
+    printf("Downloading.. %.2f%% complete.\n", (fetch->dataOffset + fetch->numBytes) * 100.0 / fetch->totalBytes);
```

Comment:
In the test harness the `Content-Length` fields are always present, so `fetch->totalBytes` is always there, but agreed that for people reading the code to use as copy-paste examples to their project, it's an omission. Added the checks to the test suite so that people won't get any surprises.